### ISSUE TITLE
runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by ialidzhikov** [kubernetes-sigs/azurefile-csi-driver#247]
**What happened**:
`azurefile-csi-driver` panics on mount operation.

**What you expected to happen**:
I would expect to run the azurefile-csi-driver with `useInstanceMetadata: true` and have mount operations working. 

**How to reproduce it**:
The Azure config is such as:
```
    cloud: AZUREPUBLICCLOUD
    location: "westeurope"
    resourceGroup: ""
    routeTableName: ""
    securityGroupName: ""
    subnetName: ""
    vnetName: ""
    primaryAvailabilitySetName: ""
    loadBalancerSku: "basic"
    cloudProviderBackoff: true
    cloudProviderBackoffRetries: 6
    cloudProviderBackoffExponent: 1.5
    cloudProviderBackoffDuration: 5
    cloudProviderBackoffJitter: 1.0
    cloudProviderRateLimit: true
    cloudProviderRateLimitQPS: 10
    cloudProviderRateLimitBucket: 100
    cloudProviderRateLimitQPSWrite: 10
    cloudProviderRateLimitBucketWrite: 100
    tenantId: ""
    subscriptionId: ""
    useInstanceMetadata: true
```

Logs of `azurefile-csi-driver`:
```
$ k -n kube-system logs csi-driver-node-file-2x86c csi-driver
E0422 11:42:43.615211       1 histogram.go:90] ---&{  cloudprovider_azure_api_request_duration_seconds Latency of an Azure API call map[] [] <nil> {0 {0 0}} {0 {0 0}} ALPHA}---
E0422 11:42:43.616642       1 histogram.go:94] ---&{  cloudprovider_azure_api_request_duration_seconds Latency of an Azure API call map[] [] <nil> {0 {0 0}} {0 {0 0}} ALPHA}---
I0422 11:42:43.716163       1 main.go:83] set up prometheus server on [::]:29615
I0422 11:42:43.716459       1 azurefile.go:106]
DRIVER INFORMATION:
-------------------
Build Date: "2020-03-20T13:38:47Z"
Compiler: gc
Driver Name: file.csi.azure.com
Driver Version: v0.5.0
Git Commit: d52c62d0391d9e21a277a0b9edfd1a7fbc3bb014
Go Version: go1.13.4
Platform: linux/amd64

Streaming logs below:
I0422 11:42:43.716488       1 azure.go:31] AZURE_CREDENTIAL_FILE env var set as /etc/kubernetes/cloudprovider/cloudprovider.conf
I0422 11:42:43.717065       1 azure_auth.go:152] Using AZUREPUBLICCLOUD environment
I0422 11:42:43.717136       1 azure.go:372] Azure cloud provider is starting without credentials
I0422 11:42:43.717150       1 azure.go:411] Azure cloudprovider using try backoff: retries=6, exponent=1.500000, duration=5, jitter=1.000000
I0422 11:42:43.816581       1 mount_linux.go:170] Cannot run systemd-run, assuming non-systemd OS
I0422 11:42:43.816603       1 mount_linux.go:171] systemd-run failed with: exit status 1
I0422 11:42:43.816619       1 mount_linux.go:172] systemd-run output: Failed to create bus connection: No such file or directory
I0422 11:42:43.816632       1 driver.go:93] Enabling controller service capability: CREATE_DELETE_VOLUME
I0422 11:42:43.816640       1 driver.go:93] Enabling controller service capability: PUBLISH_UNPUBLISH_VOLUME
I0422 11:42:43.816644       1 driver.go:93] Enabling controller service capability: CREATE_DELETE_SNAPSHOT
I0422 11:42:43.816649       1 driver.go:93] Enabling controller service capability: EXPAND_VOLUME
I0422 11:42:43.816655       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_WRITER
I0422 11:42:43.816659       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_READER_ONLY
I0422 11:42:43.816664       1 driver.go:112] Enabling volume access mode: MULTI_NODE_READER_ONLY
I0422 11:42:43.816668       1 driver.go:112] Enabling volume access mode: MULTI_NODE_SINGLE_WRITER
I0422 11:42:43.816672       1 driver.go:112] Enabling volume access mode: MULTI_NODE_MULTI_WRITER
I0422 11:42:43.816677       1 driver.go:103] Enabling node service capability: STAGE_UNSTAGE_VOLUME
I0422 11:42:43.816851       1 server.go:104] Listening for connections on address: &net.UnixAddr{Name:"//csi/csi.sock", Net:"unix"}
I0422 11:43:34.405217       1 utils.go:111] GRPC call: /csi.v1.Identity/Probe
I0422 11:43:34.405246       1 utils.go:112] GRPC request:
I0422 11:43:34.405253       1 utils.go:118] GRPC response: ready:<value:true >
I0422 11:44:02.753613       1 utils.go:111] GRPC call: /csi.v1.Node/NodeGetCapabilities
I0422 11:44:02.753635       1 utils.go:112] GRPC request:
I0422 11:44:02.753642       1 nodeserver.go:244] NodeGetCapabilities called with request {{} [] 0}
I0422 11:44:02.753652       1 utils.go:118] GRPC response: capabilities:<rpc:<type:STAGE_UNSTAGE_VOLUME > >
I0422 11:44:02.774648       1 utils.go:111] GRPC call: /csi.v1.Node/NodeStageVolume
I0422 11:44:02.774676       1 utils.go:112] GRPC request: volume_id:"#foo#" staging_target_path:"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/foo/globalmount" volume_capability:<mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:"skuName" value:"Standard_LRS" > volume_context:<key:"storage.kubernetes.io/csiProvisionerIdentity" value:"1587554147812-1847-file.csi.azure.com" >
I0422 11:44:02.774693       1 nodeserver.go:114] NodeStageVolume called with request {#foo# map[] /var/lib/kubelet/plugins/kubernetes.io/csi/pv/foo/globalmount mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER >  map[] map[skuName:Standard_LRS storage.kubernetes.io/csiProvisionerIdentity:1587554147812-1847-file.csi.azure.com] {} [] 0}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x123a4d4]

goroutine 24 [running]:
k8s.io/legacy-cloud-providers/azure.(*Cloud).GetStorageAccesskey(0xc0000c7500, 0xc0004cc181, 0x17, 0xc0002012e0, 0x19, 0x0, 0x0, 0x0, 0x0)
	/root/go/pkg/mod/k8s.io/legacy-cloud-providers@v0.0.0-20191231115903-c0243ebb3644/azure/azure_storageaccount.go:75 +0xb4
sigs.k8s.io/azurefile-csi-driver/pkg/azurefile.(*Driver).getAccountInfo(0xc0003f4090, 0xc0004cc180, 0x57, 0x0, 0xc0003c2ab0, 0x1, 0x1, 0xc0004c13c0, 0x41484f, 0x7ff212e75800, ...)
	/root/go/src/sigs.k8s.io/azurefile-csi-driver/pkg/azurefile/azurefile.go:379 +0x448
sigs.k8s.io/azurefile-csi-driver/pkg/azurefile.(*Driver).NodeStageVolume(0xc0003f4090, 0x19d0840, 0xc0003c2990, 0xc0003ba4e0, 0xc0003f4090, 0x24993cd, 0xc000366930)
	/root/go/src/sigs.k8s.io/azurefile-csi-driver/pkg/azurefile/nodeserver.go:131 +0x18a
github.com/container-storage-interface/spec/lib/go/csi._Node_NodeStageVolume_Handler.func1(0x19d0840, 0xc0003c2990, 0x16df460, 0xc0003ba4e0, 0x1, 0x1, 0x200, 0x18)
	/root/go/pkg/mod/github.com/container-storage-interface/spec@v1.1.0/lib/go/csi/csi.pb.go:5330 +0x86
sigs.k8s.io/azurefile-csi-driver/pkg/csi-common.logGRPC(0x19d0840, 0xc0003c2990, 0x16df460, 0xc0003ba4e0, 0xc000370680, 0xc0003706a0, 0xc00030da80, 0x48b578, 0x16673c0, 0xc0003c2990)
	/root/go/src/sigs.k8s.io/azurefile-csi-driver/pkg/csi-common/utils.go:114 +0x1ed
github.com/container-storage-interface/spec/lib/go/csi._Node_NodeStageVolume_Handler(0x1737700, 0xc0003f4090, 0x19d0840, 0xc0003c2990, 0xc00048a6e0, 0x182aa20, 0x19d0840, 0xc0003c2990, 0xc0000e69a0, 0x154)
	/root/go/pkg/mod/github.com/container-storage-interface/spec@v1.1.0/lib/go/csi/csi.pb.go:5332 +0x14b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004a8480, 0x19ee5e0, 0xc0000d6d80, 0xc0004c6100, 0xc0003e1860, 0x2518d20, 0x0, 0x0, 0x0)
	/root/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:971 +0x46a
google.golang.org/grpc.(*Server).handleStream(0xc0004a8480, 0x19ee5e0, 0xc0000d6d80, 0xc0004c6100, 0x0)
	/root/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:1250 +0xd16
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000046ac0, 0xc0004a8480, 0x19ee5e0, 0xc0000d6d80, 0xc0004c6100)
	/root/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:690 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/root/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:688 +0xa1
```

**Anything else we need to know?**:

**Environment**:
- CSI Driver version: `mcr.microsoft.com/k8s/csi/azurefile-csi:v0.6.0`
- Kubernetes version (use `kubectl version`): v1.18.2
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by andyzhangx** [comment]
in your logs, it said:
```
I0422 11:42:43.717136       1 azure.go:372] Azure cloud provider is starting without credentials
```

for controller component, need service principal or managed identity credential to access the cluster resource group,

**Comment by andyzhangx** [comment]
check details about how to set azure credentials for this driver: https://github.com/kubernetes-sigs/azurefile-csi-driver#prerequisite

**Comment by ialidzhikov** [comment]
`azuredisk-csi-driver` can successfully mount/unmount using `useInstanceMetadata: true` and not specifying `aadClientId` and `aadClientSecret `. So `azuredisk-csi-driver` is not requiring them, but `azurefile-csi-driver` is (at least this is what I get)?

/cc @rfranzke

**Comment by andyzhangx** [comment]
I know what you meant, this PR fixed this issue: https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/241, and on master branch, `azurefile-csi-driver` node pod could read from k8s serect to get account key, and don't need `aadClientId`, `aadClientSecret`, this change is only on master branch now.

**Comment by ialidzhikov** [comment]
So using the read-from-secret approach over `AZURE_CREDENTIAL_FILE` env var, we should be able to run the `azurefile-csi-driver` only with `useInstanceMetadata: true`, right?

**Comment by andyzhangx** [comment]
> So using the read-from-secret approach over `AZURE_CREDENTIAL_FILE` env var, we should be able to run the `azurefile-csi-driver` only with `useInstanceMetadata: true`, right?

@ialidzhikov on current master branch, we supports
 - read azure credential from secret
 - azurefile-csi-driver node pod would read storage account key from k8s secret

These two features are not relevant, so you could still use your original config(use AZURE_CREDENTIAL_FILE` env var), and master branch should work with `useInstanceMetadata: true`.

**Comment by ialidzhikov** [comment]
I am trying the current master in our setup, but I hit a RBAC related issue.

```
$ k describe pvc file-claim

  Normal   Provisioning          30s (x13 over 24m)    file.csi.azure.com_csi-driver-controller-file-9cbdf444f-zr8w2_dac561e0-e82b-4c03-a15f-a905047f4c55  External provisioner is provisioning volume for claim "default/file-claim"
  Warning  ProvisioningFailed    29s (x13 over 24m)    file.csi.azure.com_csi-driver-controller-file-9cbdf444f-zr8w2_dac561e0-e82b-4c03-a15f-a905047f4c55  failed to provision volume with StorageClass "files": rpc error: code = Internal desc = failed to store storage account key: couldn't create secret secrets is forbidden: User "system:serviceaccount:foo:default" cannot create resource "secrets" in API group "" in the namespace "default"
```

The azurefile-csi-driver Pod specifies `.spec.serviceAccount: csi-driver-node-file` and the RBAC is adapted such as:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
  name: csi-driver
rules:
- apiGroups:
  - ""
  resources:
  - secrets
  verbs:
  - create
  - list
  - update
  - get
  - patch
  - delete
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
  - list
  - update
- apiGroups:
  - ""
  resources:
  - namespaces
  verbs:
  - get
  - list
- apiGroups:
  - ""
  resources:
  - persistentvolumes
  verbs:
  - get
  - list
  - watch
  - update
  - patch
- apiGroups:
  - storage.k8s.io
  resources:
  - volumeattachments
  verbs:
  - get
  - list
  - watch
  - update
  - patch
- apiGroups:
  - policy
  - extensions
  resourceNames:
  - csi-driver-node
  resources:
  - podsecuritypolicies
  verbs:
  - use
```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
  name: csi-driver-file
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: csi-driver
subjects:
- kind: ServiceAccount
  name: csi-driver-node-file
  namespace: kube-system
```

For some reason it is trying to use ` "system:serviceaccount:foo:default"` and not the ServiceAccount specified in the Pod.

**Comment by ialidzhikov** [comment]
Hm, actually the RBAC needs to be adapted for the azuredisk-csi-driver which is currently running out-of-cluster with `--nodeid=dummy`. Currently the azuredisk-csi-driver running on the node (and responsible for mount operations) is trying to access the storage account secret but this secret is not available for this cluster as the out-of-cluster azurefile-csi-driver (responsible for volume operations) created it in different cluster (in our setup - the cluster, hosting the control plane). 
Isn't this read-from-secret change breaking the setup with 2 azurefile-csi-driver controllers - 1 running out-of-cluster and 1 running in-cluster?

**Comment by andyzhangx** [comment]
> Hm, actually the RBAC needs to be adapted for the azurefile-csi-driver which is currently running out-of-cluster with `--nodeid=dummy`. Currently the azurefile-csi-driver running on the node (and responsible for mount operations) is trying to access the storage account Secret but this secret is not available for this cluster as the out-of-cluster azurefile-csi-driver (responsible for volume operations) created it in different cluster (in our setup - the cluster, hosting the control plane).

@ialidzhikov since your azurefile-csi-driver controller and node pods are running on two different clusters, you cannot store secret in one cluster and pass it to another cluster. So in your scenario, `useInstanceMetadata: true` is not enough, node pod still needs identity(service principal or managed identity) to access that storage account.

> Isn't this read-from-secret change breaking the setup with 2 azurefile-csi-driver controllers - 1 running out-of-cluster and 1 running in-cluster?

that depends on whether the 2 azurefile-csi-driver controllers needs different identity, anyway, you could use the old way to set azure credentials, that's just another option.

**Comment by ialidzhikov** [comment]
We were missing the `--kubeconfig` flag for the out-of-cluster azurefile-csi-driver. The issue is fixed by #241.

/close
